### PR TITLE
[Testing:Forum] cypress test- display pronouns in forum

### PR DESCRIPTION
**Pull request by yanliw123**
some more cypress test for display pronouns in forum. 
1.create thread
2.verify pronouns is shown in overall forum page
3.comment on the thread, verify pronouns is shown
4.remove thread

also add code to change display pronouns option back.

### CODE REVIEW

**Review by NiaHeermance — CHANGES_REQUESTED** [review]
Looks good! I would just also include logging in as another user and seeing Joe's pronouns are there.

File: site/cypress/e2e/Cypress-UI/pronouns_change.spec.js [hunk 1]
```diff
@@ -36,16 +44,34 @@ describe('Tests cases abut changing user pronouns', () => {
...
+
         cy.get('button[aria-label="Clear pronoun input"]').click(); //clear input using trash can
         if (oldPronouns !== '') {
             cy.get('#user-pronouns-change').type(oldPronouns);
         }
-        cy.get('#edit-pronouns-submit').first().click();
 
-        //ensure pronouns changed on page
+        //change display option back
+        cy.get('#pronouns-forum-display').then(() => {
```
Comment (IDzyre) [inline site/cypress/e2e/Cypress-UI/pronouns_change.spec.js:54]: if this is only running with the student profile, would it be more concise to hard-code the resetting of the variables, because you should be able to see what the original values are ?

File: site/cypress/e2e/Cypress-UI/pronouns_change.spec.js [hunk 2]
```diff
@@ -72,4 +97,67 @@ describe('Tests cases abut changing user pronouns', () => {
...
+
+        //comment on the thread, verify pronouns is shown
+        cy.get('.create-post-head').should('contain', 'Test pronouns display');
+        cy.get('#reply_box_2').type('my pronouns is They/Them{ctrl}{enter}');
+        cy.contains('Submit Reply to All').click();
+        cy.get('.post_box').should('contain', 'my pronouns is They/Them');
+        cy.get('.post_user_pronouns').should('contain', 'They/Them');
+
+        //remove thread
+        cy.get('.first_post > .post-action-container > .delete-post-button').click();
```
Comment (IDzyre) [inline site/cypress/e2e/Cypress-UI/pronouns_change.spec.js:127]: Is there a better selector for this?

File: site/cypress/e2e/Cypress-UI/pronouns_change.spec.js [hunk 3]
```diff
@@ -72,4 +97,67 @@ describe('Tests cases abut changing user pronouns', () => {
...
+        cy.get('.create-post-head').should('contain', 'Test Anonymous thread, should not show pronouns');
+        cy.get('.thread-anon-checkbox').filter(':visible').click();
+        cy.get('#reply_box_2').type('I can not see your pronouns{ctrl}{enter}');
+        cy.contains('Submit Reply to All').click();
+        cy.get('.post_box').should('contain', 'I can not see your pronouns');
+        cy.get('.post_user_id').should('contain', 'Anonymous');
+        cy.get('.post_user_pronouns').should('not.exist');
+
+        //remove thread
+        cy.get('.first_post > .post-action-container > .delete-post-button').click();
```
Comment (IDzyre) [inline site/cypress/e2e/Cypress-UI/pronouns_change.spec.js:158]: same with better selector

**Review by IDzyre — CHANGES_REQUESTED** [review]

File: site/cypress/e2e/Cypress-UI/pronouns_change.spec.js (hunk 1 above)
Comment (yanliw123) [reply site/cypress/e2e/Cypress-UI/pronouns_change.spec.js:54]: yea, but the variables might be different or changed by different users. So In the future, if people run the cypress test locally. Their student account's variable will be changed to the original values I set, instead of their original value.

Comment (IDzyre) [reply site/cypress/e2e/Cypress-UI/pronouns_change.spec.js:54]: Will they be changing the student account pronoun for a reason on a dev VM? (besides to test something)

Comment (williamjallen) [reply site/cypress/e2e/Cypress-UI/pronouns_change.spec.js:54]: I think it makes sense to try to accept a system in any state, and then restore the system to the initial state when the test is done whenever possible.  That said, we do have a fair number of tests which rely upon the system being in a fresh state, so it's not terrible to do it that way either...

File: site/package.json [hunk 4]
```diff
@@ -71,7 +71,8 @@
     "regenerator-runtime": "^0.14.0",
     "stylelint": "^15.10.3",
     "stylelint-config-standard": "^33.0.0",
-    "typescript": "^5.0.4",
+    "ts-node": "^10.9.1",
+    "typescript": "^5.2.2",
```
Comment (williamjallen) [inline site/package.json:75]: These changes are unrelated and should be reverted.

File: site/package-lock.json
Comment (williamjallen) [inline site/package-lock.json:1]: These changes are also unrelated.

**Review by williamjallen — DISMISSED** [review]

File: site/cypress/e2e/Cypress-UI/pronouns_change.spec.js [hunk 5]
```diff
@@ -36,16 +30,18 @@ describe('Tests cases abut changing user pronouns', () => {
...
-        }
+
+        //set to false
+        cy.get('#pronouns-forum-display').uncheck();
+
+        cy.get('#edit-pronouns-submit').click();
+
+        //ensure pronouns and display option changed on page
+        cy.get('#pronouns_val').should('have.text', ' ');
+        cy.get('#display_pronouns_val').contains('False');
```
Comment (IDzyre) [inline site/cypress/e2e/Cypress-UI/pronouns_change.spec.js:44]: if you are asserting that an element contains a value, the most clear way is to do cy.get().should('contain', 'text_to_contain').

cy.get().contains('text_to_contain') will have the same output (it will fail if the element doesn't have the text), but it is for requesting the element, to then chain commands after it (like on line 78), instead of asserting that the element exists.

File: site/cypress/e2e/Cypress-UI/pronouns_change.spec.js [hunk 6]
```diff
@@ -72,4 +67,79 @@ describe('Tests cases abut changing user pronouns', () => {
...
+        //verify pronouns exist and remove thread
+        cy.contains('Test pronouns display').find('.post_user_pronouns').should('contain','They/Them');
+        cy.contains('Test pronouns display').find('.post_user_pronouns').click();
+        cy.get('.post_user_pronouns').should('contain', 'They/Them');
+        cy.get('.first_post').find('.fa-trash').click();
+        cy.contains('Test pronouns display').should('not.exist');
+
+        //verify pronouns do not exist (post thread Anonymously) and remove thread
+        cy.contains('Test Anonymous thread, should not show pronouns').find('.post_user_id').should('contain','Anonymous');
+        cy.contains('Test Anonymous thread, should not show pronouns').find('.post_user_pronouns').should('not.be.exist');
```
Comment (IDzyre) [inline site/cypress/e2e/Cypress-UI/pronouns_change.spec.js:137]:
```suggestion
        cy.contains('Test Anonymous thread, should not show pronouns').find('.post_user_pronouns').should('not.exist');
```

File: site/cypress/e2e/Cypress-UI/pronouns_change.spec.js [hunk 7]
```diff
@@ -72,4 +67,79 @@ describe('Tests cases abut changing user pronouns', () => {
...
+
+        //verify pronouns is shown in overall fourm page
+        cy.get('#nav-sidebar-forum').click();
+        cy.get('.thread-list-item').should('contain', 'Test pronouns display');
+        cy.contains('Test pronouns display').find('.post_user_pronouns').should('contain','They/Them');
+        cy.contains('Test pronouns display').find('.post_user_pronouns').click();
+
+        //comment on the thread, verify pronouns is shown
+        cy.get('.create-post-head').should('contain', 'Test pronouns display');
+        cy.get('#reply_box_2').type('my pronouns is They/Them{ctrl}{enter}');
```
Comment (IDzyre) [inline site/cypress/e2e/Cypress-UI/pronouns_change.spec.js:91]:
```suggestion
        cy.get('#reply_box_2').type('my pronouns are They/Them{ctrl}{enter}');
```

**Review by IDzyre — CHANGES_REQUESTED** [review]
I like how everything is being tested in one 'it' statement. This may be a 'little harder to read' for newer developers than splitting the test up, however its best practice to have all tests be self-sufficient, and not require others tests before to setup.

**Review by NiaHeermance — APPROVED** [review]
Test now checks if pronouns are these as another user. Looks good!

File: site/cypress/e2e/Cypress-UI/pronouns_change.spec.js [hunk 8]
```diff
@@ -12,18 +9,15 @@ describe('Tests cases abut changing user pronouns', () => {
...
-        //type in new pronouns
+        //type in new pronouns and check display in forum option
         cy.get('button[aria-label="Clear pronoun input"]').click(); //clear input using trash can
         cy.get('@e').type('They/Them');
+        cy.get('#pronouns-forum-display').check();
         cy.get('#edit-pronouns-submit').click();
 
-        //ensure pronouns changed on page
+        //ensure pronouns and display option changed on page
         cy.get('#pronouns_val').contains('They/Them');
```
Comment (IDzyre) [inline site/cypress/e2e/Cypress-UI/pronouns_change.spec.js:19]: Same comment about the should('contain', 'text') for this line and the one after.

**Review by IDzyre — CHANGES_REQUESTED** [review]
One more comment about cypress assertions, then it looks good.

**Review by IDzyre — APPROVED** [review]

**Comment by bmcutler** [comment]
@yanliw123 @IDzyre 
Can one of you please resolve the merge conflicts on this PR?
Thanks!
